File: Assn5.cpp
#include <iostream>
#include "string"
#include <vector> 
#include <array>
#include <queue> 
#include <algorithm>
#include <iterator>
using namespace std;

vector<string> arrList;
vector<string> evalExpr;


//generates the combinations of the character array from the length of the input number
void generatePermutations(vector<char> set, string prefix, int n, int k)
{
	if (k == 0) {
		if (prefix.at(0) != '@') 
		{
			arrList.insert(arrList.end(), prefix);
		}
		return;
	}

	for (int i = 0; i < n; ++i) 
	{
		string newPrefix = prefix + set[i];
		generatePermutations(set, newPrefix, n, k - 1);
	}
}


//helper function to allow the filtering of the @ sign in the string containing both numbers and operators
void replaceHelper(string & targetString, char keyChar)
{
	targetString.erase(remove(targetString.begin(), targetString.end(), '@'), targetString.end());
}

//inserts the numbers inbetween the list of operators: from ++-@+@@++ to 1+2+3-45+67+8+9
void insertNumbers(vector<string> arr, string digits)
{
	for (size_t i = 0; i < arr.size(); i++) {
		string str = "";
		for (size_t j = 0; j < digits.length(); j++) {
			if (j == 0) {
				str = str + "0";
			}

			str = str + (arr[i]).at(j);
			str = str + (digits).at(j);
		}
		replaceHelper(str, '@');
		evalExpr.insert(evalExpr.end(), str);
	}
}

//evaluate method that takes in a string math expression and return an integer number
int evaluate(string s)
{
	int expTotal = 0; //total of the passed in string that will be returned
	bool adding = true; //bool used to determine if the evaluator is currently adding or subtracting from the total
	for (int i = 0; i < s.length(); i = i + 0){ //looping through the expression in the form 0+1+2+3-45+67+8+9
		string tempNum = ""; //temporary string used to build numbers more than one digit such as 45
		if (s.at(i) == '+'){ //checking to see if the next number to be evaluated will be added or subtracted
			adding = true;
			i++; //iterating to the next value in the list
		}
		else if(s.at(i) == '-'){ //checking if there is a minus
			adding = false;
			i++;
		}
		while (i < s.length() && s.at(i) != '+' && s.at(i) != '-'){ //beggining to build the number string until it either hits an operator or reaches the end of the string
			tempNum = tempNum + s.at(i); //building the string of numbers
			i++;
		}
		if (adding == true){ //once the string of numbers has been made such as 45, the number is either added or subtracted from the grand total depending on the bool flag
			expTotal = expTotal + stoi(tempNum);
		}
		else{
			expTotal = expTotal - stoi(tempNum);
		}
	}

	return expTotal; //once all of the numbers have been added or subtracted the total is returned and evaluated in the main
}

int main(int argc, char * argv[])
{
	//taking in the arguments from the command line 
	string nums = argv[1];
	string sum = argv[2];
	

	vector<char> opSet;
	opSet.push_back('+');
	opSet.push_back('-');
	opSet.push_back('@');


	size_t k = nums.length();
	int count = 0;

	generatePermutations(opSet, "", opSet.size(), k);
	insertNumbers(arrList, nums);

	// Evaluates a String into an Expression
	for (size_t i = 0; i < evalExpr.size(); i++)
	{
		int total = evaluate(evalExpr[i]);// gets the expression total
		if (total == stoi(sum))
		{// checks if total is equal to the sum

			if(evalExpr[i].at(1) == '+')
				cout << evalExpr[i].substr(2) << " = " << sum << endl;
			else
				cout << evalExpr[i].substr(1) << " = " << sum << endl; // output expresstion and total
			count++;// increments the number of found solutions
		}
	}
	cout << "There are " << count << " possible solutions." << endl;

}

File: Assn5.java
import java.util.*;

public class Assn5{ 
	

    public static ArrayList<String> arrList = new ArrayList<String>(); //an array consisting of all the combinations of operators used with the length of the input number 
    public static ArrayList<String> evalExpr = new ArrayList<String>(); //the resulting array once the numbers have been placed inline with the operators

    public static void main(String[] args){
        char[] opSet = {'+', '-', '@'}; //character array to be passed into the generator
        String nums = args[0];// user input String number in terminal
        int sum = Integer.parseInt(args[1]);// user input String sum in terminal
        int count = 0;// keeps track of how many solutions were found
        int k = nums.length(); //storing the length of the input string of numbers, ex. 123456789 = 9
        
        generatePermutations(opSet, "", opSet.length, k); //creates every possible equation for the number and outputs it in an array
        insertNumbers(arrList, nums); // takes the resulting array of operators from the previous line and inserts numbers inbetween the operators

        // loops through the list of evaluable expressions ex. (1+2+3+4-5+67+89) and outputs if the expression equals the target number
        for(int i = 0; i < evalExpr.size(); i++){
            int total = evaluate(evalExpr.get(i));// gets the expression total
            if(total == sum){// checks if expression total is equal to the target number passed in
                if (evalExpr.get(i).charAt(1) == '+')
                    System.out.println(evalExpr.get(i).substring(2) + " = " + sum);
                else
                    System.out.println(evalExpr.get(i).substring(1) + " = " + sum);// output expresstion and total
                count++;// increments the number of found solutions
            }
        }
        System.out.println("There are " + count + " possible solutions.");// output number of solutions 
    } 
    

	//method used to fill an array list with a list of every combination of operators that are passed in as
	/*
	 * set: a set of character operators such as +, -, @
	 * prefix: a reacurring string that is built up from the elements of the set
	 * n: the amount of elements within the character array
	 * k: the amount of characters within the target number ex. 123456789 = 9
	 */
    public static void generatePermutations(char[] set, String prefix, int n, int k){ 
        if (k == 0){ //checking to see if the very first run through
            // @ represents 2 or more numbers that are together
            if(prefix.charAt(0) != '@'){ //adding the very first character of the expression as long as it is not an @ sign giving us equations with only + or - in the front
                arrList.add(prefix);
            }
            return; 
        } 
		//iterates through each character in the operator and recursively builds the string of operators resulting in something that looks like: ++-+-+@+
        for (int i = 0; i < n; ++i){ 
            String newPrefix = prefix + set[i];  
            generatePermutations(set, newPrefix, n, k - 1);  
        } 
    } 
    
	//takes the resulting array from generatePermutations in the form (+-+-++@+) and inserts number to create a list of evaluatable strings such as (1+2-3+4-5+6+78+9)
    public static void insertNumbers(ArrayList<String> arr, String digits){
        for(int i = 0; i < arr.size(); i++){ 	//looping through each string in the array to modify
            String str = ""; //temp string that will be used to build the evaluatable expression
            for(int j = 0; j < digits.length(); j++){ //now looping through one string within the list and beggining to insert numbers in it
                if(j == 0){ //to make parsing the expression easier a 0 is added in front of each one
                    str = str + "0"; 
                }
                str = str + (arr.get(i)).charAt(j); //because the expression always follows the pattern (num)(op)(num)(op)...(num)
                str = str + (digits).charAt(j); 	//the string is built following the added zero, by grabbing an op then a number and so on until the whole expression is built 
            }
            str = str.replace("@", ""); //in order to create the scenarios that look like this: (1+2-3+4-5+6+78+9) an @ sign was used to combine digits together ex. (1+2-3+4-5+6+7@8+9), thus it is removed at the end
            evalExpr.add(str); //adding the created expression to the list of evaluatable expressions
        }
    }
    
    public static int evaluate(String s){
        int i;// used to loop through each char in String
        int total = 0;// keeps track of expression total

        // gets integer from expression
        if(!s.contains("+") && !s.contains("-")){
            return Integer.parseInt(s);
        }

        // gets operator from expression
        for(i = s.length() - 1; i >= 0; i--){
            if (s.charAt(i) == '+' || s.charAt(i) == '-'){
                break;
            }
        }
    
        String x = s.substring(0, i);// first integer
        String y = s.substring(i + 1, s.length());// second integer
    
        // updates total based on the operator
        switch(s.charAt(i)){
            case '+':// addition
                total = evaluate(x) + evaluate(y);
                break;
            case '-':// subtraction
                total = evaluate(x) - evaluate(y);
                break;
        }
        return total;// returns expression total
    }    

}